cs/src/Connections/RelayTunnelConnector.cs

```diff
@@ -0,0 +1,222 @@
+                stream = await this.relayClient.CreateSessionStreamAsync(cancellation);
+                await this.relayClient.ConfigureSessionAsync(stream, isReconnect, cancellation);
+
+                stream = null;
+                disconnectReason = SshDisconnectReason.None;
+                return;
+            }
+            catch (UnauthorizedAccessException uaex) // Tunnel access token validation failed.
+            {
+                await RefreshTunnelAccessTokenAsync(uaex);
+            }
+            catch (SshReconnectException srex)
+            {
+                errorDescription = srex.Message;
+                disconnectReason = SshDisconnectReason.ProtocolError;
+                exception = srex;
+                isDelayNeeded = false;
+                isReconnect = false;
+            }
+            catch (SshConnectionException scex) when (scex.DisconnectReason == SshDisconnectReason.ConnectionLost)
+            {
+                // Recoverable
+                errorDescription = scex.Message;
+            }
+            catch (SshConnectionException scex)
+            {
+                // All other SSH connection exceptions are not recoverable.
+                disconnectReason = scex.DisconnectReason != SshDisconnectReason.None ? scex.DisconnectReason : SshDisconnectReason.ByApplication;
+                throw exception = new TunnelConnectionException($"Failed to start tunnel SSH session: {scex.Message}", scex);
+            }
+            catch (WebSocketException wse)
+            {
+                if (wse.WebSocketErrorCode == WebSocketError.UnsupportedProtocol)
+                {
+                    throw exception = new InvalidOperationException("Unsupported web socket sub-protocol.", wse);
+                }
+
+                if (wse.WebSocketErrorCode == WebSocketError.NotAWebSocket)
+                {
+                    int? statusCode = wse.Data.Contains("HttpStatusCode") ? (int)wse.Data["HttpStatusCode"]! : null;
+                    switch (statusCode)
+                    {
+                        case 401:
+                            // Unauthorized error may happen when the tunnel access token is no longer valid, e.g. expired.
+                            // Try refreshing it.
+                            await RefreshTunnelAccessTokenAsync(wse);
+                            break;
+
+                        case 403:
+                            throw exception = new UnauthorizedAccessException(
+                                $"Forbidden (403). Provide a fresh tunnel access token with '{this.relayClient.TunnelAccessScope}' scope.", 
+                                wse);
+
+                        case 404:
+                            throw exception = new TunnelConnectionException($"The tunnel or port is not found (404).", wse);
+
+                        case 429:
+                            errorDescription = "Rate limit exceeded (429). Too many requests in a given amount of time.";
```

Comment:
Probably there should be some delay before retrying after a 429.

I'm currently implementing the server-side code that may return 429. The first limit will be a maximum number of concurrent client connections (not a connection rate). So after a 429 it won't succeed on retry unless some other client has disconnected in the meantime.

Reply:
Oh, I see `isDelayNeeded` will be true at this point, so it's good.